README.md
# go-exercises-worker

[![GoDoc](https://godoc.org/github.com/thewizardplusplus/go-exercises-worker?status.svg)](https://godoc.org/github.com/thewizardplusplus/go-exercises-worker)
[![Go Report Card](https://goreportcard.com/badge/github.com/thewizardplusplus/go-exercises-worker)](https://goreportcard.com/report/github.com/thewizardplusplus/go-exercises-worker)
[![Build Status](https://app.travis-ci.com/thewizardplusplus/go-exercises-worker.svg?branch=master)](https://app.travis-ci.com/thewizardplusplus/go-exercises-worker)
[![codecov](https://codecov.io/gh/thewizardplusplus/go-exercises-worker/branch/master/graph/badge.svg)](https://codecov.io/gh/thewizardplusplus/go-exercises-worker)

Background worker of the service for solving programming exercises.

## Features

- interaction with queues:
  - common properties:
    - automatic declaring of the used queues;
    - passing of a message data in JSON;
  - operations:
    - consuming of the solutions:
      - concurrent handling;
      - once requeue the solution on failure;
    - producing of the solution results:
      - generating of the custom message ID;
- solution runners:
  - solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
    - loading of the allowed imports from an outer configuration;
    - canceling by a timeout:
      - code compiling;
      - code execution;
- tools:
  - script for generating the allowed import configuration.

## Installation

Prepare the directory:

```
$ mkdir --parents "$(go env GOPATH)/src/github.com/thewizardplusplus/"
$ cd "$(go env GOPATH)/src/github.com/thewizardplusplus/"
```

Clone this repository:

```
$ git clone https://github.com/thewizardplusplus/go-exercises-worker.git
$ cd go-exercises-worker
```

Install dependencies with the [dep](https://golang.github.io/dep/) tool:

```
$ dep ensure -vendor-only
```

Build the project:

```
$ go install ./...
```

## Usage

```
$ go-exercises-backend
```

Environment variables:

- `ALLOWED_IMPORT_CONFIG` &mdash; path to the allowed import config (default: `./configs/allowed_imports.json`);
- `RUNNING_TIMEOUT` &mdash; maximal duration of solution running (default: `10s`);
- message broker:
  - `MESSAGE_BROKER_ADDRESS` &mdash; [RabbitMQ](https://www.rabbitmq.com/) connection URI (default: `amqp://rabbitmq:rabbitmq@localhost:5672`);
  - `MESSAGE_BROKER_BUFFER_SIZE` &mdash; [RabbitMQ](https://www.rabbitmq.com/) channel capacity (default: `1000`);
- `SOLUTION_CONSUMER_CONCURRENCY` &mdash; amount of solution consumer threads (default: `1000`).

## API Description

API description in the [AsyncAPI](https://www.asyncapi.com/) format: [docs/async_api.yaml](docs/async_api.yaml).

## License

The MIT License (MIT)

Copyright &copy; 2021 thewizardplusplus


CHANGE_LOG.md
# Change Log

## [v1.1.6](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.6) (2022-02-27)

Change the port for displaying the API description and fix the dependencies.

- change the port for displaying the API description;
- fix the dependencies:
  - specify the version of the `@asyncapi/html-template` package;
  - update the dependencies:
    - update the `github.com/thewizardplusplus/go-rabbitmq-utils` package to v1.1.4;
    - update the `asyncapi/generator` image to v1.9.0.

## [v1.1.5](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.5) (2021-08-11)

Describe releases of the project.

- describe for releases:
  - features;
  - change log.

## [v1.1.4](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.4) (2021-08-05)

Add the API description in the [AsyncAPI](https://www.asyncapi.com/) format.

- API description in the [AsyncAPI](https://www.asyncapi.com/) format:
  - describe the API:
    - for consuming of the solutions;
    - for producing of the solution results;
  - add the Docker Compose configuration:
    - for generating of the API description;
    - for displaying of the API description.

## [v1.1.3](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.3) (2021-08-04)

Install the [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) tool in the Travis CI configuration and add the general integration tests.

- rename the used environment variables;
- [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) tool:
  - specify the version of the tool in the Docker configuration;
  - install the tool in the Travis CI configuration;
- add the general integration tests;
- fix and complete the badges in the README.md file.

## [v1.1.2](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.2) (2021-06-26)

Cover with unit tests the loading of the allowed imports from an outer configuration, and add the script for generating such configuration.

- solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
  - loading of the allowed imports from an outer configuration:
    - cover the code with unit tests;
- add the script for generating the allowed import configuration.

### Features

- interaction with queues:
  - common properties:
    - automatic declaring of the used queues;
    - passing of a message data in JSON;
  - operations:
    - consuming of the solutions:
      - concurrent handling;
      - once requeue the solution on failure;
    - producing of the solution results:
      - generating of the custom message ID;
- solution runners:
  - solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
    - loading of the allowed imports from an outer configuration;
    - canceling by a timeout:
      - code compiling;
      - code execution;
- tools:
  - script for generating the allowed import configuration.

## [v1.1.1](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1.1) (2021-06-21)

Generating of the custom message ID on producing of the solution results, and canceling the code compiling by a timeout in the solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package.

- dependencies:
  - update the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package;
  - use the [github.com/thewizardplusplus/go-rabbitmq-utils](https://github.com/thewizardplusplus/go-rabbitmq-utils) package;
- producing of the solution results:
  - generating of the custom message ID;
  - cover the code with unit tests;
- solution runners:
  - remove the dummy solution runner;
  - solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
    - canceling the code compiling by a timeout;
    - cover the code with unit tests.

### Features

- interaction with queues:
  - common properties:
    - automatic declaring of the used queues;
    - passing of a message data in JSON;
  - operations:
    - consuming of the solutions:
      - concurrent handling;
      - once requeue the solution on failure;
    - producing of the solution results:
      - generating of the custom message ID;
- solution runners:
  - solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
    - loading of the allowed imports from an outer configuration;
    - canceling by a timeout:
      - code compiling;
      - code execution.

## [v1.1](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.1) (2021-04-06)

Canceling the code execution by a timeout in the solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package.

- solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
  - canceling the code execution by a timeout.

### Features

- interaction with queues:
  - common properties:
    - automatic declaring of the used queues;
    - passing of a message data in JSON;
  - operations:
    - consuming of the solutions:
      - concurrent handling;
      - once requeue the solution on failure;
    - producing of the solution results;
- solution runners:
  - dummy solution runner (returns the current timestamp as the result);
  - solution runner based on the [github.com/thewizardplusplus/go-code-runner](https://github.com/thewizardplusplus/go-code-runner) package:
    - loading of the allowed imports from an outer configuration;
    - canceling the code execution by a timeout.

## [v1.0](https://github.com/thewizardplusplus/go-exercises-worker/tree/v1.0) (2021-03-29)


cmd/go-exercises-worker/main.go
package main

import (
	"log"
	"os"
	"os/signal"
	"time"

	"github.com/caarlos0/env"
	"github.com/go-log/log/print"
	"github.com/thewizardplusplus/go-exercises-worker/gateways/config"
	"github.com/thewizardplusplus/go-exercises-worker/gateways/queues"
	"github.com/thewizardplusplus/go-exercises-worker/gateways/runners"
	rabbitmqutils "github.com/thewizardplusplus/go-rabbitmq-utils"
)

type options struct {
	AllowedImportConfig string        `env:"ALLOWED_IMPORT_CONFIG" envDefault:"./configs/allowed_imports.json"` // nolint: lll
	RunningTimeout      time.Duration `env:"RUNNING_TIMEOUT" envDefault:"10s"`
	MessageBroker       struct {
		Address string `env:"MESSAGE_BROKER_ADDRESS" envDefault:"amqp://rabbitmq:rabbitmq@localhost:5672"` // nolint: lll
	}
	SolutionConsumer struct {
		BufferSize  int `env:"SOLUTION_CONSUMER_BUFFER_SIZE" envDefault:"1000"`
		Concurrency int `env:"SOLUTION_CONSUMER_CONCURRENCY" envDefault:"1000"`
	}
}

const (
	solutionQueueName       = "solution_queue"
	solutionResultQueueName = "solution_result_queue"
)

func main() {
	logger := log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lmicroseconds)

	var options options
	if err := env.Parse(&options); err != nil {
		logger.Fatalf("[error] unable to parse the options: %v", err)
	}

	allowedImports, err := config.LoadAllowedImports(options.AllowedImportConfig)
	if err != nil {
		logger.Fatalf("[error] unable to load the allowed imports: %v", err)
	}

	messageBrokerClient, err := rabbitmqutils.NewClient(
		options.MessageBroker.Address,
		rabbitmqutils.WithMaximalQueueSize(options.SolutionConsumer.BufferSize),
		rabbitmqutils.WithQueues([]string{
			solutionQueueName,
			solutionResultQueueName,
		}),
	)
	if err != nil {
		logger.Fatalf("[error] unable to create the message broker client: %v", err)
	}
	defer func() {
		if err := messageBrokerClient.Close(); err != nil {
			logger.Fatalf("[error] unable to close the message broker client: %v", err)
		}
	}()

	solutionConsumer, err := rabbitmqutils.NewMessageConsumer(
		messageBrokerClient,
		solutionQueueName,
		rabbitmqutils.Acknowledger{
			MessageHandling: rabbitmqutils.TwiceMessageHandling,
			MessageHandler: rabbitmqutils.JSONMessageHandler{
				MessageHandler: queues.SolutionHandler{
					SolutionResultQueueName: solutionResultQueueName,
					SolutionRunner: runners.SolutionRunner{
						AllowedImports: allowedImports,
						RunningTimeout: options.RunningTimeout,
						Logger:         print.New(logger),
					},
					MessagePublisher: messageBrokerClient,
				},
			},
			Logger: print.New(logger),
		},
	)
	if err != nil {
		logger.Fatalf("[error] unable to create the solution consumer: %v", err)
	}
	go solutionConsumer.StartConcurrently(options.SolutionConsumer.Concurrency)
	defer func() {
		if err := solutionConsumer.Stop(); err != nil {
			logger.Fatalf("[error] unable to stop the solution consumer: %v", err)
		}
	}()

	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt)
	<-interrupt
}
